smart_open/s3.py

```diff
@@ -1061,7 +1062,14 @@ def _list_bucket(
         else:
             for c in content:
                 key = c['Key']
-                if accept_key(key):
+                # Enable accept_key() to optionally take `content` as a second argument.
```

Comment:
I think it would be better to probe the accept_key function once, outside of the loops, and then call it as appropriate.

For the probe mechanism, I think checking the number of parameters explicitly is better than relying on TypeError, because it's possible (although unlikely) that the user's code can raise TypeError.

Reply:
OK, I agree. This makes the code a bit less readable, IMO, but it is probably significantly more performant. How about this approach?:
```
            if accept_key.__code__.co_argcount == 1:
                chooser = lambda k,c: accept_key(k)
            else:
                chooser = accept_key
            for c in content:
                key = c['Key']
                if chooser(key, c):
                    yield key
```

If this seems acceptable to you, I'll put it in and address your other comments.

Reply:
I would do this:

```
def _wrap_for_compatibility(accept_key):
    def inner(key, content):
        return accept_key(key)
    return inner

#
# Handle legacy accept_key functions: they only accepted a single param.
#
if accept_key.__code__.co_argcount == 1:
    accept_key = _wrap_for_compatibility(accept_key)

for c in content:
    if accept_key(c['Key'], c):
        yield c['Key]
```

In such cases, it's better to avoid lambda. Lambda makes things a bit more magical here, and it's also [against PEP8](https://stackoverflow.com/questions/25010167/e731-do-not-assign-a-lambda-expression-use-a-def).